### PR TITLE
Namelist I/O

### PR DESCRIPTION
**Pull request by vdonaldson**

### CODE REVIEW

File: flang/lib/Lower/IO.cpp [hunk 1]
```diff
@@ -196,6 +196,144 @@ static void makeNextConditionalOn(Fortran::lower::FirOpBuilder &builder,
...
+                           bool wantResult = false) {
+    auto global =
+        createStringLiteral(builder, loc, symbol.name().ToString() + '\0');
+    return wantResult
+               ? builder.createConvert(loc, charPtrTy, fir::getBase(global))
+               : mlir::Value{};
+  };
+  stringAddress(symbol);
+  for (const Fortran::semantics::Symbol &s : details.objects())
+    stringAddress(s);
```
Comment (jeanPerier) [inline flang/lib/Lower/IO.cpp:251]: If the goal of `stringAddress` gymnastic  is to order the global creations according to their dependencies, wouldn't it be simpler to put the character globals in a smallVector an later looping over it  ?

```
llvm::SmallVector<mlir::Value> nameAddresses;
for (const Fortran::semantics::Symbol &s : details.objects()) {
  auto global =  createStringLiteral(builder, loc, s.name().ToString() + '\0');
  nameAddresses.push_back(builder.createConvert(loc, charPtrTy, fir::getBase(global)));
}
auto listFunc = [&](Fortran::lower::FirOpBuilder &builder) {
  // ...
 for (auto name : nameAddresses)
   list = builder.create<fir::InsertValueOp>(loc, listTy, list, name, idx);
 // ...
}
```

File: flang/lib/Lower/IO.cpp [hunk 2]
```diff
@@ -196,6 +196,144 @@ static void makeNextConditionalOn(Fortran::lower::FirOpBuilder &builder,
...
+      list = builder.create<fir::InsertValueOp>(loc, listTy, list, name, idx);
+    }
+    builder.create<fir::HasValueOp>(loc, list);
+  };
+  auto list =
+      builder.createGlobal(loc, listTy, listName, false, listFunc, linkOnce);
+  auto groupFunc = [&](Fortran::lower::FirOpBuilder &builder) {
+    mlir::Value group = builder.create<fir::UndefOp>(loc, groupTy);
+    n = 0;
+    auto groupName = stringAddress(symbol, true);
```
Comment (jeanPerier) [inline flang/lib/Lower/IO.cpp:272]: Same here, would the following work ?:
```
auto groupName = builder.createConvert(loc, charPtrTy, fir::getBase(createStringLiteral(...));
auto groupFunc = [&](Fortran::lower::FirOpBuilder &builder) {
   ....
  group = builder.create<fir::InsertValueOp>(loc, groupTy, group, groupName,
}
```

File: flang/lib/Lower/IO.cpp [hunk 3]
```diff
@@ -196,6 +196,144 @@ static void makeNextConditionalOn(Fortran::lower::FirOpBuilder &builder,
...
+///         const char *name; // NUL-terminated lower-case
+///         const Descriptor &descriptor;
+///       };
+///       const char *groupName; // NUL-terminated lower-case
+///       std::size_t items;
+///       const Item *item; // in original declaration order
+///     };
+///
+/// Return the Item array and the NamelistGroup data structure.
+static std::pair<fir::GlobalOp, fir::GlobalOp>
```
Comment (jeanPerier) [inline flang/lib/Lower/IO.cpp:216]: Is this really OK for these structures to be globals modified before IO calls using them ?
I guess re-entrency is not much of an issue given it I do not think the IO statement can contain a call that itself uses the same namelist, but wouldn't it be an issue when mixing with OpenMP/OpenACC ?

Does nvfortran support namelist IO in parallel contexts ?

File: flang/lib/Lower/IO.cpp [hunk 4]
```diff
@@ -196,6 +196,144 @@ static void makeNextConditionalOn(Fortran::lower::FirOpBuilder &builder,
...
+  mlir::Value groupAddr = builder.create<fir::AddrOfOp>(
+      loc, namelistGroup.resultType(), namelistGroup.getSymbol());
+  mlir::Value list = builder.create<fir::AddrOfOp>(
+      loc, namelistList.resultType(), namelistList.getSymbol());
+  auto one = builder.createIntegerConstant(loc, int32Ty, 1);
+  int n = 0;
+  for (const Fortran::semantics::Symbol &s : details.objects()) {
+    auto nextIdx = builder.createIntegerConstant(loc, idxTy, n++);
+    llvm::SmallVector<mlir::Value, 2> idx = {nextIdx, one};
+    // const std::optional<Fortran::evaluate::Expr<Fortran::evaluate::SomeType>>
```
Comment (jeanPerier) [inline flang/lib/Lower/IO.cpp:315]: left-over ?

**Review by jeanPerier — COMMENTED** [review]
Good job getting the Namelist struct type working ! My main comments is that I am not sure it's OK to use and modify globals as storage for these structs. I also think it should be possible to avoid the `fir.ref<fir.ptr<>>` type with inserts (though I know you probably did try already).

File: flang/lib/Lower/IO.cpp (hunk 3 above)
Comment (vdonaldson) [reply flang/lib/Lower/IO.cpp:216]: I've switched to regenerating all information for each use of a namelist group.  The net result is that only constant name strings are reused.

File: flang/lib/Lower/IO.cpp [hunk 5]
```diff
@@ -196,6 +196,144 @@ static void makeNextConditionalOn(Fortran::lower::FirOpBuilder &builder,
...
+  }
+  const auto &details = symbol.get<Fortran::semantics::NamelistDetails>();
+  auto loc = converter.getCurrentLocation();
+  auto *context = builder.getContext();
+  auto linkOnce = builder.createLinkOnceLinkage();
+  auto idxTy = builder.getIndexType();
+  auto sizeTy = builder.getIntegerType(8 * sizeof(std::size_t));
+  auto charPtrTy = fir::PointerType::get(builder.getIntegerType(8));
+  auto descPtrTy =
+      fir::PointerType::get(fir::BoxType::get(mlir::NoneType::get(context)));
```
[the remark this answers is not in the document]
Comment (vdonaldson) [reply flang/lib/Lower/IO.cpp:235]: I've switched to `InsertOpValue`, and removed uses of `fir::PointerType`.

File: flang/lib/Lower/IO.cpp (hunk 1 above)
Comment (vdonaldson) [reply flang/lib/Lower/IO.cpp:251]: `stringAddress` has been simplified.

File: flang/lib/Lower/IO.cpp (hunk 2 above)
Comment (vdonaldson) [reply flang/lib/Lower/IO.cpp:272]: All namelist component accesses now use `fir::InsertValueOp`.

File: flang/lib/Lower/IO.cpp (hunk 4 above)
Comment (vdonaldson) [reply flang/lib/Lower/IO.cpp:315]: Yes, thanks.

File: flang/lib/Lower/IO.cpp [hunk 6]
```diff
@@ -1395,8 +1524,8 @@ void genBeginCallArguments(llvm::SmallVectorImpl<mlir::Value> &ioArgs,
                            Fortran::lower::AbstractConverter &converter,
                            mlir::Location loc, const A &stmt,
                            mlir::FunctionType ioFuncTy, bool isFormatted,
-                           bool isList, bool isIntern, bool isOtherIntern,
-                           bool isAsynch, bool isNml,
+                           bool isListOrNml, bool isIntern,
+                           [[maybe_unused]] bool isOtherIntern, bool isAsynch,
```
[the remark this answers is not in the document]
Comment (vdonaldson) [reply flang/lib/Lower/IO.cpp:1528]: Unfortunately, this `[[maybe_unused]]` is needed for some instantiations of this template function.

File: flang/lib/Optimizer/Dialect/FIRType.cpp [hunk 7]
```diff
@@ -358,7 +358,13 @@ BoxProcType::verify(llvm::function_ref<mlir::InFlightDiagnostic()> emitError,
...
 }
 
-static bool canBePointerOrHeapElementType(mlir::Type eleTy) {
+static bool cannotBePointerElementType(mlir::Type eleTy) {
+  return eleTy.isa<BoxProcType, ShapeType, ShapeShiftType,
+                   SliceType, FieldType, LenType, HeapType, PointerType,
+                   ReferenceType, TypeDescType>();
+}
+
+static bool cannotBeHeapElementType(mlir::Type eleTy) {
```
Comment (jeanPerier) [inline flang/lib/Optimizer/Dialect/FIRType.cpp:367]: Unless you still need those, I think it might be best to revert the changes in these file.

File: flang/lib/Lower/IO.cpp [hunk 8]
```diff
@@ -196,6 +196,87 @@ static void makeNextConditionalOn(Fortran::lower::FirOpBuilder &builder,
...
+    auto global =
+        createStringLiteral(builder, loc, symbol.name().ToString() + '\0');
+    return builder.createConvert(loc, charRefTy, fir::getBase(global));
+  };
+  makeNextConditionalOn(builder, loc, checkResult, ok);
+  auto zero = builder.createIntegerConstant(loc, idxTy, 0);
+  auto one = builder.createIntegerConstant(loc, idxTy, 1);
+  auto two = builder.createIntegerConstant(loc, idxTy, 2);
+  auto listAddr = builder.create<fir::AllocaOp>(loc, listTy);
+  mlir::Value list = builder.create<fir::LoadOp>(loc, listAddr);
```
Comment (jeanPerier) [inline flang/lib/Lower/IO.cpp:241]: Given you are not relying on some data already being set in the storage like you did with the names in the global, it would probably be best to start with an undefined (I think this will avoid LLVM optimizers having to think whether or not the storage can change between the load and final store , if it cannot be sure, it would probably generate another temp from the load, modify it, and copy it back).
```
mlir::Value list = builder.create<fir::UndefOp>(loc, fir::dyn_cast_ptrEleTy(listAddr.getType()));
```

**Review by jeanPerier — COMMENTED** [review]
Val that looks great to me. A few more nits on your update.

There is one corner case you might not have considered regarding modules (and probably also internal procedures). The objects in the namelist might not be instantiated if only the namelist symbol is imported in the current scope.


Could you try if the following program works:
```
module m
 integer :: x, y
 namelist /nm/ x, y
end module

subroutine  foo()
  use m, only: nm
  write(*, nm)
end subroutine
```

If it hit an unknown symbol assert, you might need to add something to make sure the namelist objects are added to the PFT symbols (that analyze() is called on the namelist objects).

I think you may have to consider the `NamelistDetails` case right after https://github.com/flang-compiler/f18-llvm-project/blob/4cf0cc3803b3e572af441033553f96c49cb747ef/flang/lib/Lower/PFTBuilder.cpp#L1315

File: flang/lib/Optimizer/Dialect/FIRType.cpp (hunk 7 above)
Comment (vdonaldson) [reply flang/lib/Optimizer/Dialect/FIRType.cpp:367]: Done; thanks

File: flang/lib/Lower/IO.cpp (hunk 8 above)
Comment (vdonaldson) [reply flang/lib/Lower/IO.cpp:241]: done; thanks

**Review by jeanPerier — APPROVED** [review]

File: flang/lib/Lower/ConvertVariable.cpp [hunk 9]
```diff
@@ -335,6 +335,14 @@ static void instantiateGlobal(Fortran::lower::AbstractConverter &converter,
                                               global.getSymbol());
   Fortran::lower::StatementContext stmtCtx;
   mapSymbolAttributes(converter, var, symMap, stmtCtx, addrOf);
+  if (sym.test(Fortran::semantics::Symbol::Flag::InNamelist)) {
```
[the remark this answers is not in the document]
Comment (vdonaldson) [reply flang/lib/Lower/ConvertVariable.cpp:338]: The motivation for this alias code is to avoid duplicate `fir.address_of` ops.  This test shows the issue:

```
    module mmm
      namelist /nnn/ jjj
    end module

      use mmm
      jjj = 17
      write(*, nnn)
    end
```

In function `main` there are two symbols for `jjj`.  One is a direct local symbol, which is an alias for the other ultimate `jjj` symbol from the module.  The purpose of the `instantiateGlobal` code is to map both of these symbols to the same `fir.address` op.  Without this `instantiateGlobal` code, the generated high level FIR is:

```
func @_QQmain() {
  %0 = fir.alloca !fir.box<i32> {uniq_name = ""}
  %1 = fir.address_of(@_QMmmmEjjj) : !fir.ref<i32>
  %2 = fir.address_of(@_QMmmmEjjj) : !fir.ref<i32>
  %c17_i32 = constant 17 : i32
  fir.store %c17_i32 to %1 : !fir.ref<i32>
  . . .
  %11 = fir.embox %2 : (!fir.ref<i32>) -> !fir.box<i32>
```

Some optimization pass collapses these two `fir.address_of` ops into a single instance.  The `instantiateGlobal` code avoids this duplication up front.

For a namelist group defined in a module, it is probably safe to let optimization clean up this duplication because the namelist object is a global.  Another possibility though is an uplevel access to a local namelist
group object in a containing subprogram where the code would otherwise incorrectly generate addresses for two distinct objects, rather than duplicate addresses for a single object.  Since we don't yet have uplevel referencing, I'm not sure if this is actually a correctness issue or not.

[the remark this answers is not in the document]
Comment (vdonaldson) [reply flang/lib/Lower/ConvertVariable.cpp:338]: A little more thought suggests that this `instantiateGlobal` instance of the code is probably optional, but there might need to also be similar code in `instantiateLocal`.

[the remark this answers is not in the document]
Comment (jeanPerier) [reply flang/lib/Lower/ConvertVariable.cpp:338]: I do not think the same situation can happen in `instantiateLocal`, but I guess similar thing can happen with common/equivalenced. However, the consequences of this is only a double instantiation (once for the HostAssociated symbol, and once for the ultimate one) that CSE can indeed easily re-group as you mentioned. I would tend to not add special handling/complexity in lowering (from a pedantic way, your code still duplicates the specification expression evaluation, which are also simple expressions that gets CSEed).

If we really wants to take it at some point, I would rather that we find a way to re-group the HostAssociated/Ultimate symbols into a single variables during the PFT analysis rather than to have ad-hoc handling. 

Also, this case only happen with Namelist defined in modules, and you will most likely change the way this case is handled soon anyway (making the NamelistGroup a global), so this symbol will not be used anymore.

[the remark this answers is not in the document]
Comment (vdonaldson) [reply flang/lib/Lower/ConvertVariable.cpp:338]: Ok, I'll remove this code.  We can revisit this if we eventually find cases where having two different symbols for a namelist object causes generation of two objects, rather than just two object addresses.

**Review by mleair — COMMENTED** [review]
LGTM

File: flang/lib/Optimizer/Support/InternalNames.cpp [hunk 10]
```diff
@@ -205,6 +205,15 @@ fir::NameUniquer::doVariable(llvm::ArrayRef<llvm::StringRef> modules,
   return result.append(toLower(name));
 }
 
+std::string
+fir::NameUniquer::doNamelistGroup(llvm::ArrayRef<llvm::StringRef> modules,
+                                  llvm::Optional<llvm::StringRef> host,
+                                  llvm::StringRef name) {
+  std::string result = prefix();
+  result.append(doModulesHost(modules, host)).append("NG");
```
Comment (schweitzpgi) [inline flang/lib/Optimizer/Support/InternalNames.cpp:213]: Just use one capital letter here to be consistent with all other cases of unique names,

File: flang/lib/Lower/IO.cpp [hunk 11]
```diff
@@ -196,6 +197,183 @@ static void makeNextConditionalOn(Fortran::lower::FirOpBuilder &builder,
...
+///     class NamelistGroup {
+///     public:
+///       struct Item {
+///         const char *name; // NUL-terminated lower-case
+///         const Descriptor &descriptor;
+///       };
+///       const char *groupName; // NUL-terminated lower-case
+///       std::size_t items;
+///       const Item *item; // in original declaration order
+///     };
```
Comment (schweitzpgi) [inline flang/lib/Lower/IO.cpp:212]: While I personally find this to be a well-written description, I'm also wondering about including a snippet of code like this because it just isn't something the F18 project really does (or does much). I suppose the argument against it is that there is no way for someone changing `namelist.h` to know that this comment is here.

**Review by schweitzpgi — APPROVED** [review]
Nice work, Val. Looks great. Thank you.

File: flang/lib/Optimizer/Support/InternalNames.cpp (hunk 10 above)
Comment (vdonaldson) [reply flang/lib/Optimizer/Support/InternalNames.cpp:213]: It looks like there are other 2-letter cases in this file, but I can change to a single letter if that is preferred.  Any suggestions?  (Hmm, why "E" instead of "V" for variable?  "E" -> entity??)

```
doModules - "M" "S"
doModulesHost - "F"
doKind - "K" "N"
doCommonBlock - "B"
doBlockData - "L"
doConstant - "EC"
doDispatchTable - "DT"
doGenerated - "Q"
doIntrinsicTypeDescriptor - "C"
doProcedure - "P"
doType - "T"
doTypeDescriptor - "CT"
doVariable - "E"

doNamelistGroup - "NG" ??
```

File: flang/lib/Lower/IO.cpp (hunk 11 above)
Comment (vdonaldson) [reply flang/lib/Lower/IO.cpp:212]: That's a valid point.  I'll change this to just reference the definition, rather than copying it.

File: flang/lib/Optimizer/Support/InternalNames.cpp (hunk 10 above)
Comment (vdonaldson) [reply flang/lib/Optimizer/Support/InternalNames.cpp:213]: I'll switch to "G".  This choice is user-visible in object code, but can be changed at least up to the time of a first official release.
